jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from contextlib import contextmanager
+from typing import Any, Dict, Callable
+
+from jax import config, core
+from jax.interpreters.xla import DeviceArray
+
+np_impl: Dict[core.Primitive, Callable[..., Any]] = {}
+is_tracing = False
```

Comment:
Global state isn't a good idea. It isn't threadsafe, and moreover if it's not in `core.thread_local_state.trace_state` then it won't be copied when we need to create thunks for delayed computation.

Let's delete this `is_tracing` global.

Reply:
Agree, is removed.